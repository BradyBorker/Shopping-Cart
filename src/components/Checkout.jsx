import styles from "../styles/checkout.module.css"
import genStyles from "../styles/general.module.css"

const Checkout = ({ cart, setCart }) => {
    // id, title, price, category, description, image
    const cartObjs = Object.keys(cart).map((key) => (
        cart[key]
    ))

    const removeProduct = (obj) => {
        if (Number(obj.count) === 1) {
            setCart((currentCart) => {
                // eslint-disable-next-line no-unused-vars
                const { [obj.product.id]: omitted, ...rest } = currentCart;
                return rest
            })
        } else {
            setCart((currentCart) => {
                const newCount = Number(currentCart[`${obj.product.id}`].count) - 1;
                return { ...cart, [`${obj.product.id}`]: { count: newCount, product: obj.product } };
            })
        }
    }

    const getTotalCost = () => {
        let total = 0;
        cartObjs.forEach((obj) => {
            total += parseFloat(obj.product.price) * Number(obj.count)
        })

        return total;
    }

    return (
        <>
            <h2 className={genStyles.centerText}>Shopping Cart</h2>
            <main>
                {cartObjs.map((obj, index) => (
                    <section key={index} className={styles.cartRow}>
                        <img src={obj.product.image} alt={obj.product.description} className={styles.img} />
                        <div className={styles.text}>{obj.product.title}</div>
                        <div className={styles.text}>${obj.product.price}</div>
                        <div className={styles.text}>Count: {obj.count}</div>
                        <button type='button' className={styles.button} onClick={() => removeProduct(obj)}>Remove</button>
                    </section>
                ))}
                <section className={styles.total}>
                    {cartObjs.length === 0 ? 'Cart is Empty!' : `Total Cost: ${getTotalCost().toFixed(2)}`}
                </section>
            </main>
        </>
    )
};

export default Checkout;
