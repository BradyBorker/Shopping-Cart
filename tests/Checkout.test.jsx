import { render, screen } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import Checkout from '../src/components/Checkout';
import userEvent from '@testing-library/user-event';

const cart = { '0': { product: { id: 0, image: '', title: 'product1', price: 1 }, count: 1 } }

describe('Checkout component', () => {
    it('renders the Checkout component', () => {
        const { container } = render(<Checkout cart={{}} setCart={() => null} />)
        expect(container).toMatchSnapshot();
    })

    it('renders when cart is empty', () => {
        render(<Checkout cart={{}} setCart={() => null} />)

        const heading = screen.getByRole('heading', { name: 'Shopping Cart' });
        expect(heading).toBeInTheDocument();

        const main = screen.getByRole('main');
        expect(main.childElementCount).toBe(1);

        const emptyMessage = screen.getByText('Cart is Empty!');
        expect(emptyMessage).toBeInTheDocument();
    })

    it('renders when products are in the cart', () => {
        render(<Checkout cart={cart} setCart={() => null} />)

        const main = screen.getByRole('main');
        expect(main.childElementCount).toBe(2);

        const totalCost = screen.getByText(/Total Cost: \$\d+\.?\d*/i);
        expect(totalCost).toBeInTheDocument();
    })

    it('calls removeProduct when the remove button is clicked', async () => {
        const user = userEvent.setup();
        const setCart = vi.fn(() => {
            cart['0'].count = cart['0'].count - 1
        })

        render(<Checkout cart={cart} setCart={setCart} />);

        const button = screen.getByRole('button', { name: 'Remove' });
        await user.click(button)

        expect(setCart).toBeCalledTimes(1);
    })
})