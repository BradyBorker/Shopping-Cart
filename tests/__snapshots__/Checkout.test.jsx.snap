// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkout component > renders the Checkout component 1`] = `
<div>
  <h2
    class="_centerText_be7884"
  >
    Shopping Cart
  </h2>
  <main>
    <section
      class="_total_fa8560"
    >
      Cart is Empty!
    </section>
  </main>
</div>
`;
